src/pbft/pbft.h

```diff
@@ -147,42 +168,62 @@ namespace pbft
         0,
         pbft_network.get(),
         &message_receiver_base);
+      LOG_INFO_FMT("PBFT setup for self with id: {}", local_id);
 
       pbft_config->set_service_mem(mem + used_bytes);
+      pbft_network->set_receiver(message_receiver_base);
 
-      LOG_INFO_FMT("Setting up client proxy");
-      static auto client_proxy =
-        std::make_unique<ClientProxy<CallerId, void>>(*message_receiver_base);
+      Byz_start_replica();
+
+      LOG_INFO_FMT("PBFT setting up client proxy");
+      client_proxy =
+        std::make_unique<ClientProxy<kv::TxHistory::RequestID, void>>(
+          *message_receiver_base);
 
       auto cb = [](Reply* m, void* ctx) {
-        auto cp = static_cast<ClientProxy<CallerId, void>*>(ctx);
+        auto cp =
+          static_cast<ClientProxy<kv::TxHistory::RequestID, void>*>(ctx);
         cp->recv_reply(m);
       };
 
       message_receiver_base->register_reply_handler(cb, client_proxy.get());
 
       on_request = [&](kv::TxHistory::CallbackArgs args) {
-        auto caller = std::get<0>(args.id);
-        auto session = std::get<1>(args.id);
-        auto jsonrpc_id = std::get<2>(args.id);
-
         auto total_req_size = pbft_config->message_size() + args.data.size();
 
         uint8_t request_buffer[total_req_size];
         pbft_config->fill_request(
-          request_buffer, total_req_size, args.data, jsonrpc_id);
-
-        Time t = ITimer::current_time();
-
-        client_proxy->send_request(
-          t,
+          request_buffer,
+          total_req_size,
+          args.data,
+          args.actor,
+          args.caller_id);
+
+        auto rep_cb = [&](
+                        void* owner,
+                        kv::TxHistory::RequestID caller_rid,
+                        int status,
+                        uint8_t* reply,
+                        size_t len) {
+          LOG_DEBUG_FMT("PBFT reply callback for {}", caller_rid);
+
+          return rpcsessions.reply_async(
+            std::get<1>(caller_rid), {reply, reply + len});
+        };
+
+        LOG_DEBUG_FMT("PBFT sending request {}", args.rid);
+        return client_proxy->send_request(
+          args.rid,
           request_buffer,
           sizeof(request_buffer),
-          nullptr,
+          rep_cb,
           client_proxy.get());
       };
     }
 
+    // TODO(#PBFT): PBFT consensus should implement the following functions to
```

Comment:
We need a single abstract Consensus interface, let's talk about this on Monday.

Reply:
As discussed, we will address this a a future PR.